log: skip on unknown nested mixing instead of stopping the logging

eliransin opened the pull request:
The logging function for an exception type will stop if a nested
exception change is not mixed in with a type derived from
std::exception. The consequence of this is that if there are more
nested exceptions in the chain they will not be printed out which
means that we can lose viable debugging information.
This commit fixes this by adding a catch handler that will skip the
unknown part of the chain and will try to continue to the next part of
the exception chain which is the std::nested_exception member.

It also adds explicit indication in the logging where there is no
information on an object - for example when it is not derived from
std::exception.

Tests for the exception logging are added for all of the different
cases.

Fixes #874

tgrabiec commented on src/util/log.cc:
```diff
@@ -553,6 +553,23 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
             }
+        } catch (const std::nested_exception& ne) {
+           // if we got here, it means that the nested_exception is not mixed in
+           // with an object of type std::exception this it can mean one of two
```
` this it` -> `. It`

tgrabiec commented on src/util/log.cc:
```diff
@@ -553,6 +553,23 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
+           //    throw and not in the context of another catch with std::throw_with_nested
+           // 2. Somewhere in the code there is a throw of a type which doesn't derive from
+           //    std::exception and then it is rethrown using std::throw_with_nested.
+           out<< ": " << "unknown nested type";
```
Missing space after `out`.

tgrabiec commented on src/util/log.cc:
```diff
@@ -553,6 +553,23 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
+           if (ne.nested_ptr()) {
+                // skip over the mixed in object and continue recursively to the
+                // unfold the contained exception itself
+                out <<  ": " << ne.nested_ptr();
```
Double space after first <<

tgrabiec commented on src/util/log.cc:
```diff
@@ -553,6 +553,23 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
+                out <<  ": " << ne.nested_ptr();
+           } else {
+               // there is no contained exception
+               out<< ": " << "empty contained exception";
```
Missing space after `out`.

tgrabiec reviewed (changes requested):
I only have some cosmetic issues

nyh commented on src/util/log.cc:
```diff
@@ -553,6 +553,23 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
+           //    throw and not in the context of another catch with std::throw_with_nested
+           // 2. Somewhere in the code there is a throw of a type which doesn't derive from
+           //    std::exception and then it is rethrown using std::throw_with_nested.
+           out << ": " << "unknown nested type";
```
Why is this an "unknown nested type"?  When we print ne.nested_ptr() below, doesn't it know it's own type? We have std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) which does exactly that.

eliransin replied:
The resulting `std::nested_exception` type when generated properly with `std::throw_with_nested(T object)` is not a pure `std::nested_exception` but a **mix of two types** - a nested_type that *contains* the current exception, and the type
contained in the object. something like `class x : public T, public nested_exception {...` 
We can't account for every T in our code so we also account for objects derived from `std::exception` and then we print 
the what.
The "unknown nested type" print, indicates that we don't know the **whole** nested type, or more specifically, what T is and hence we can't print it.
About the test - fair enough, I will add a test instead the code in the commit message.

nyh replied:
Hmm, I think I see now what you meant about the "unknown nested type". Ok.

nyh reviewed (changes requested):
I have one question on the "unknown nested type" message below, but what bothers me even more is the commit message. The whole point of this patch is to improve the way we print nested exceptions, so why does the example in the commit message use a different print_exception function? It should use "<<" and show how the code below really prints nested exceptions! This really confused me, when initially I didn't understand how come the output looks different from what the code outputs.

Also I didn't understand why this code is hard to test... Looks very easy to test! Just throw the exceptions exactly like in the commit message, "<<" the resulting exception into an std::stringstream, and compare the string to what you expect. Sounds easy to me, and better than putting code in the commit message (which, as I said above, doesn't even use the operator<< you added).

eliransin commented:
@nyh thanks for insisting on the tests, made me polish the logging code a little bit more :slightly_smiling_face: 
Changes from previous version:
1. Removed some redundancy - the logging was already able to handle empty pointers so no need to handle separately,
moreover it will make a more consistent log.
2. Add an explicit logging of "no more information" so it is easy to deduce the exact flow of this function given a log line.
It is an edge case anyway so if we see this printout we should probably debug the exception throwing code anyway.
3. Added tests to cover all of the cases, old and new.

nyh commented on src/util/log.cc:
```diff
@@ -553,8 +553,17 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
+            out << ": " << "unknown nested type: " << ne.nested_ptr();
         } catch(...) {
-            // no extra info
+            out << " (" << "no extra info" << ")";
```
Nitpick: this new "no extra info" message is unrelated to the issue of nested exceptions (you have an example in your test below), and I wonder if it's helpful... Basically the operator<< already printed the exception's type, but indeed doesn't know how to print any more information about its value. But I don't have strong feelings either way, so I think this is ok.

eliransin replied:
I think this helps for completeness of the log message format. In the examples it can be seen the it makes the message more regular or uniform, using an example from https://github.com/scylladb/seastar/pull/875#discussion_r587290842 :
```
unknown_obj (no extra info)
std::runtime_error (Exception Level 0)
```
If anything, I would have added this `(no extra info)` also to the `unknown nested type` remark because it stands out.
On the other hand we should never see either of them if our exception throwing code is well written so I am not sure we
should devote a fix for that.
This is my opinion of course and it is totally subjective :slightly_smiling_face:

eliransin commented on tests/unit/exception_logging_test.cc:
```diff
@@ -0,0 +1,142 @@
+
+// This function generates the expected logging output string of an exception
+// thrown by the  exception generator function with a specific output.
+std::string exception_generator_str(uint32_t test_instance,int nesting_level) {
```
I understand that point, for completeness, here are the strings that are produces by the run of this generator test:
```
unknown_obj (no extra info)
std::runtime_error (Exception Level 0)
std::_Nested_exception<unknown_obj>: unknown nested type: unknown_obj (no extra info)
std::_Nested_exception<std::runtime_error> (Exception Level 1): unknown_obj (no extra info)
std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
std::_Nested_exception<std::runtime_error> (Exception Level 1): std::runtime_error (Exception Level 0)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<unknown_obj>: unknown nested type: unknown_obj (no extra info)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<unknown_obj>: unknown nested type: unknown_obj (no extra info)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<std::runtime_error> (Exception Level 1): unknown_obj (no extra info)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<std::runtime_error> (Exception Level 1): unknown_obj (no extra info)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<std::runtime_error> (Exception Level 1): std::runtime_error (Exception Level 0)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<std::runtime_error> (Exception Level 1): std::runtime_error (Exception Level 0)
```

nyh replied:
Thanks. I already committed this series, so if we want to improve it will need to be a followup patch.
In these examples you can see what I asked about the string "unknown exception type"... Check out: 
```
std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
```
What does "unknown nested type" refer to? The message said it's std::nested_exception<unknown_obj> so we know it's an unknown_obj, and, as the next part of the message says - also an std::runtime_error. So what has "unknown type"?

eliransin replied:
Now I see the point, maybe I should change it to `non exception derived type (no extra info)`.
I can also just skip this entirely.
I am inclining towards that it is not that important but I could add another patch if it will bother someone.

nyh replied:
If you really want that "no extra info" on unknown exception types, you should - I think - at least be consistent and add it here like in non-nested exceptions. I.e., the above message will become:
```
std::_Nested_exception<unknown_obj> (no extra info): std::runtime_error (Exception Level 0)
```
i.e., no "unknown type" message (which feels wrong to me, the type is known), no "non exception derived type", just a "no extra info" exactly like in the other catch(...).

I still don't know what this "no extra info" message adds or clarifies, though. But at least it's not misleading like (I think) "unknown type" is misleading.
